src/TextObjects/Tag.ts

```diff
@@ -0,0 +1,138 @@
+import { TextDocument, Position, Range } from 'vscode'
+import {TextObject} from './TextObject';
+
+type Tag = { name: string; type: 'close' | 'open'; start: number; end: number };
+type MatchedTag = {
+    tag: string;
+    openingTagStart: number;
+    openingTagEnd: number;
+    closingTagStart: number;
+    closingTagEnd: number;
+};
+
+export class TextObjectTag extends TextObject {
+
+    TAG_REGEX = /\<(\/)?([^\>\<\s\/]+)(?:[^\>\<]*?)(\/)?\>/g
+    OPEN_FORWARD_SLASH = 1;
+    TAG_NAME = 2;
+    CLOSE_FORWARD_SLASH = 3
+    openStart: number | undefined;
+    openEnd: number | undefined;
+    closeStart: number | undefined;
+    closeEnd: number | undefined;
+    tagsParsed: boolean = false;
+
+    static byTag(args: {isInclusive: boolean}): TextObject {
+        const obj = new TextObjectTag();
+        obj.isInclusive = args.isInclusive;
+        return obj;
+    }
+
+    findStartRange(document: TextDocument, anchor: Position): Range | null {
+        this.parseTags(document, anchor);
+
+        if (this.openStart !== undefined && this.openEnd !== undefined) {
+            return new Range(
+                document.positionAt(this.openStart),
+                document.positionAt(this.openEnd)
+            )
+        }
+        return null;
+    }   
+
+    findEndRange(document: TextDocument, anchor: Position): Range | null {
+        this.parseTags(document, anchor);
+
+        if (this.closeStart !== undefined && this.closeEnd !== undefined) {
+            return new Range(
+                document.positionAt(this.closeStart),
+                document.positionAt(this.closeEnd)
+            );
+        }
+        return null;
+    }   
+
+    parseTags(document: TextDocument, anchor: Position) {
+        // Only need to parse tags if we haven't already
+        if (this.tagsParsed) {
+            return;
+        }
+
+        let match;
+        let tags: Tag[] = [];
+        while ((match = this.TAG_REGEX.exec(document.getText())) !== null) {
```

Comment:
We should avoid using `getText` without a range. This would freeze the editor if the document is too large.

VSCode recently added a feature to mirror cursor on matching tags. Seems like it's using the language server:
https://github.com/microsoft/vscode/blob/9e1d730cf2950297d7a85fc66be89411d7b8bcd9/extensions/html-language-features/client/src/htmlMain.ts#L134-L139
This will only support HTML, but maybe that's what we want anyway?

Or we can just write a more performant matching logic, but some edge cases such as `<br />` or `<div name="<div>">got you</div>` can be difficult.

Reply:
It would be good to support JSX and any similar variants if possible - not sure if using the language server would limit that?

Reply:
Yes, not to mention the other various templating languages and file extensions. Some I use often are `*.blade.php` or `*.antlers.html` for example. 

I am not a huge fan of parsing the whole document, but admittedly this idea and the regex was borrowed from the other vim extension. I tried doing it other ways but ran into various issues that required having the sets of matching end tags (which my original attempts did not).

Reply:
Make sense. We can challenge this later.

@austenc Would you mind adding a comment here such as `// TODO: Potential performance issue with getText()`?

Reply:
Done! Thanks for the reminder 👍